Mathlib/LinearAlgebra/Matrix/BilinearForm.lean

```diff
@@ -0,0 +1,623 @@
+/-
+Copyright (c) 2020 Anne Baanen. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Anne Baanen, Kexing Ying
+
+! This file was ported from Lean 3 source module linear_algebra.matrix.bilinear_form
+! leanprover-community/mathlib commit 075b3f7d19b9da85a0b54b3e33055a74fc388dec
+! Please do not edit these lines, except to modify the commit id
+! if you have ported upstream changes.
+-/
+import Mathlib.LinearAlgebra.Matrix.Basis
+import Mathlib.LinearAlgebra.Matrix.Nondegenerate
+import Mathlib.LinearAlgebra.Matrix.NonsingularInverse
+import Mathlib.LinearAlgebra.Matrix.ToLinearEquiv
+import Mathlib.LinearAlgebra.BilinearForm
+import Mathlib.LinearAlgebra.Matrix.SesquilinearForm
+
+/-!
+# Bilinear form
+
+This file defines the conversion between bilinear forms and matrices.
+
+## Main definitions
+
+ * `Matrix.toBilin` given a basis define a bilinear form
+ * `Matrix.toBilin'` define the bilinear form on `n → R`
+ * `BilinForm.toMatrix`: calculate the matrix coefficients of a bilinear form
+ * `BilinForm.toMatrix'`: calculate the matrix coefficients of a bilinear form on `n → R`
+
+## Notations
+
+In this file we use the following type variables:
+ - `M`, `M'`, ... are modules over the semiring `R`,
+ - `M₁`, `M₁'`, ... are modules over the ring `R₁`,
+ - `M₂`, `M₂'`, ... are modules over the commutative semiring `R₂`,
+ - `M₃`, `M₃'`, ... are modules over the commutative ring `R₃`,
+ - `V`, ... is a vector space over the field `K`.
+
+## Tags
+
+bilinear_form, matrix, basis
+
+-/
+
+
+variable {R : Type _} {M : Type _} [Semiring R] [AddCommMonoid M] [Module R M]
+
+variable {R₁ : Type _} {M₁ : Type _} [Ring R₁] [AddCommGroup M₁] [Module R₁ M₁]
+
+variable {R₂ : Type _} {M₂ : Type _} [CommSemiring R₂] [AddCommMonoid M₂] [Module R₂ M₂]
+
+variable {R₃ : Type _} {M₃ : Type _} [CommRing R₃] [AddCommGroup M₃] [Module R₃ M₃]
+
+variable {V : Type _} {K : Type _} [Field K] [AddCommGroup V] [Module K V]
+
+variable {B : BilinForm R M} {B₁ : BilinForm R₁ M₁} {B₂ : BilinForm R₂ M₂}
+
+section Matrix
+
+variable {n o : Type _}
+
+open BigOperators
+
+open BilinForm Finset LinearMap Matrix
+
+open Matrix
+
+/- ./././Mathport/Syntax/Translate/Expr.lean:107:6: warning: expanding binder group (i j) -/
```

Comment:
Is this line produced by mathport? Similarly elsewhere.

Reply:
Yes, I recall the best practice was to keep these warnings so we can debug mathport, but I'd be happy to remove them of course.

Reply:
Apparently I remembered wrong: https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/unsupported.3A.20ambiguous.20notation/near/338596270

```suggestion
```